rmw_fastrtps_cpp/src/rmw_publisher.cpp

```diff
@@ -101,8 +100,18 @@ rmw_create_publisher(
       static_cast<void *>(&msg),
       nullptr);
     if (RMW_RET_OK != rmw_ret) {
-      rmw_fastrtps_shared_cpp::__rmw_destroy_publisher(
+      rmw_error_state_t error_state = *rmw_get_error_state();
+      rmw_reset_error();
+      static_cast<void>(common_context->graph_cache.dissociate_writer(
+        info->publisher_gid, common_context->gid, node->name, node->namespace_));
+      rmw_ret = rmw_fastrtps_shared_cpp::__rmw_destroy_publisher(
         eprosima_fastrtps_identifier, node, publisher);
+      if (RMW_RET_OK != rmw_ret) {
```

Comment:
I don't like having to do this to ensure the first error that occurs is the one that pops up and that no subsequent error goes unnoticed.

If we had a way to tell `rcutils` that we're in error handling or finalization/cleanup mode so that it doesn't complain about error states being overwritten (and prints the new one to `stderr`, or aborts the program, or whatever), this code would simpler (and a bit more neat). What do reviewers think?

Reply:
That'd be simpler, yeah, but I think we would be trading usability in the process. To the point it makes me wonder what 's the point of returning an generic error code vs. not returning anything.

Reply:
I agree that this is a bit unfortunate.  The big problem is that you have to be careful on *every* error path to do the checking and not overwrite the error.  It will always be a game of whack-a-mole (or we have to have really good tests).

> That'd be simpler, yeah, but I think we would be trading usability in the process. To the point it makes me wonder what 's the point of returning an generic error code vs. not returning anything.

To be fair, the generic error code at least distinguishes between "succeeded" and "failed".  The question on whether the individual errors are worthwhile comes down to whether the calling code can do anything reasonable in response to the error.  If that caller can do something specific in response to some errors, then it makes sense to have individual errors.  If the caller can never do anything about the situation, the generic error is enough.

The problem in my mind with the `rcutils` flag is that it is not generic enough; you also need that flag at the rcl layer, rmw layer, and potentially the rclcpp/rclpy layers as well.

Overall, I think the error situation isn't perfect.  But I also think it is a large topic, and we don't necessarily have to solve it here.  I'd say that we continue with this PR assuming the current situation.  We can open up a separate issue and potentially have a meeting where we talk about the broader issue of errors.  Does that sound reasonable?

Reply:
> To be fair, the generic error code at least distinguishes between "succeeded" and "failed".

True, but we don't need specific return codes for that.

> The problem in my mind with the rcutils flag is that it is not generic enough; you also need that flag at the rcl layer, rmw layer, and potentially the rclcpp/rclpy layers as well.

Why? It's the same flag everywhere. It doesn't have to be a flag either, it's just that building anything bigger than this seems overkill.

> We can open up a separate issue and potentially have a meeting where we talk about the broader issue of errors. Does that sound reasonable?

Fair enough. Let's roll as-is.

Reply:
> Why? It's the same flag everywhere. It doesn't have to be a flag either, it's just that building anything bigger than this seems overkill.

I'm thinking of this situation:

```
int rcl_do_thing() {
  rcl_subthing_init();  // Succeeds
  if (rcutils_do_something() != OK) {
    // Failed, rcutils_do_something() set error code/message
    rcl_subthing_cleanup();  // Also failed, reset error code/message
    return NOT_OK;
  }
}
```

But its possible that I don't have enough context on how the error handling works.

Reply:
So, all layers use the same thread local error state and API that's defined in `rcutils` but through aliases. In your example, `rcl_do_thing` would enter this error handling mode for the extent of that `if` block statement. In a way, it's a poor man's try-catch block. 

But I see now that arbitrarily nested error handling may occur (a try-catch in the catch block of a try-catch in the catch block of...). We would have to make it reentrant and a boolean flag isn't enough. An integer count could be used instead.